rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,414 @@
+- The obvious question that might come to your mind is that what happens to the content after a breakpoint. For eg: after say `960px` the mobile typography scale would be picked up by the system but there are mobile devices who have resolutions ranging from `320px`, `480px`, `760px`, `960px` so won't the mobile typography scale look too small on say `760px` screen width device? Well, that's the reason we have kept the units of typography to be `rems` so it can adjust to these use cases based on resolutions, browsers font size, etc. This will also handle use cases for high resolution mobile screens for eg: iPhone 12. You can [read more about that in our RFC about units for layout and typography](https://github.com/razorpay/blade/blob/master/rfcs/2021-02-19-units-for-typography-and-layout.md#what-will-work-for-us)
+
+### On Consumer Code Side
+
+Taking the [approaches discussed above](#what-would-work-for-us?)
+
+**1. Components that can handle Responsiveness Internally**
+
+For this to work, on the consumer side nothing has to be done explicitly, as far as the consumers have wrapped their app inside `<BladeProvider><ConsumerApp></BladeProvider>`, the components which will handle responsiveness internally will automatically listen to the screen size changes and scale automatically based on the available space
+
+**2. Components that can give some flexibility to handle Responsiveness to consumers**
+
+Consider, if we want to implement a layout similar to the below mockup
+<img alt="Adaptive Layout with flex direction" src="./images/responsive-adaptive-layout/adaptive-layout-with-flex-direction.png">
+
+<br/>
+
+Here's how the consumer will create a layout like this in code(_Note: the actual API might differ_)
+
+```jsx
+import { Stack, Card, Title } from '@razorpay/blade/components';
+
+const Dashboard = () => (
+  <Stack direction={{ s: 'column', l: 'row' }}>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+  </Stack>
+);
+
+export default Dashboard;
+```
+
+**3. Components for Adaptive layout(rendering components conditionally)**
+
+- Consider, if we want to implement a layout similar to the below mockup where we have 2 different search inputs for different screens of the same web app.
+  <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-search-field.png">
+
+  <br/>
+
+  Here's how the consumer will create a layout like this in code(_Note: the actual API might differ_).
+
+  ```jsx
+  import { LargeSearchInput, SmallSearchInput } from '@razorpay/blade/components';
+
+  const SearchScreen = () => (isMobile ? <SmallSearchInput /> : <LargeSearchInput />);
```

Comment:
**Point 1**

This looks fine but I'm more concerned that, A naive user of a component library shouldn't care about codesplitting the components with this much granularity. 

They should just be able to say `<SearchInput />` and the input should handle the isMobile logic internally. 


**Point 2**

Second point is we didn't to mention the internal implementation, which in this case i believe is very important. Let's replace the `<SearchInput />` with something more complex component say `<Select />`

Open ended questions: 

- How will the Select's core business logic be shared between both SmallSelectInput & LargeSelectInput? We certainly don't want to repeat the same code in both places. We need some sort of `BaseSelect` component or even [hooks based renderless approach](https://survivejs.com/blog/renderlesskit-interview/) for this, where all the internal logic should be in `useSelectComponent` hook so that building both variants of the component just becomes a work of adding few different styling in most cases thus adding different rendering logic becomes easy. 

- Are there any significant changes on the behaviour of the component? (Ideally it shouldn't)

Reply:
Point 1 is similar to what @chaitanyadeorukhkar asked [here](https://github.com/razorpay/blade/pull/416#issuecomment-1049802978)

Point 2 is out of scope of this RFC. That will be covered in code sharing RFC which will be coming soon. That is not just limited to share logic for similar components but also across platforms(react web and react-native)

Reply:
finally, [answered here](https://github.com/razorpay/blade/pull/416#issuecomment-1056790157)